modules/gapi/src/backends/fluid/gfluidcore.cpp

```diff
@@ -843,6 +843,283 @@ static void run_arithm_s(DST out[], const SRC in[], int width, int chan,
+                                              const v_int32& c2, const v_int32& c3,
+                                              const v_int32& c4, const v_int32& c5,
+                                              const v_int32& c6)
+{
+    constexpr int nlanes = static_cast<int>(v_uint16::nlanes);
+    vx_store(out_ptr, v_pack_u(c1, c2));
+    vx_store(out_ptr + nlanes, v_pack_u(c3, c4));
+    vx_store(out_ptr + 2*nlanes, v_pack_u(c5, c6));
+}
+
+template<typename T>
+CV_ALWAYS_INLINE int absdiffc_simd_c3_impl(const T in[], T out[],
+                                           const v_float32& s1, const v_float32& s2,
+                                           const v_float32& s3, const int length)
+{
+    static_assert((std::is_same<T, ushort>::value) || (std::is_same<T, short>::value),
+                  "This templated overload is only for short or ushort type combinations.");
+
+    constexpr int nlanes = (std::is_same<T, ushort>::value) ? static_cast<int>(v_uint16::nlanes):
+                                                              static_cast<int>(v_int16::nlanes);
+
+    if (length < 3 * nlanes)
+        return 0;
+
+    int x = 0;
+    for (;;)
+    {
+        for (; x <= length - 3 * nlanes; x += 3 * nlanes)
+        {
+            v_float32 a1 = v_load_f32(in + x);
+            v_float32 a2 = v_load_f32(in + x + nlanes / 2);
+            v_float32 a3 = v_load_f32(in + x + nlanes);
+            v_float32 a4 = v_load_f32(in + x + 3 * nlanes / 2);
+            v_float32 a5 = v_load_f32(in + x + 2 * nlanes);
+            v_float32 a6 = v_load_f32(in + x + 5 * nlanes / 2);
+
+            absdiffc_short_store_c3(&out[x], v_round(v_absdiff(a1, s1)),
+                                             v_round(v_absdiff(a2, s2)),
+                                             v_round(v_absdiff(a3, s3)),
+                                             v_round(v_absdiff(a4, s1)),
+                                             v_round(v_absdiff(a5, s2)),
+                                             v_round(v_absdiff(a6, s3)));
+        }
+
+        if (x < length && (in != out))
+        {
+            x = length - 3 * nlanes;
+            continue;  // process one more time (unaligned tail)
+        }
+        break;
+    }
+    return x;
+}
+
+template<>
+CV_ALWAYS_INLINE int absdiffc_simd_c3_impl<uchar>(const uchar in[], uchar out[],
+                                                  const v_float32& s1, const v_float32& s2,
+                                                  const v_float32& s3, const int length)
+{
+    constexpr int nlanes = static_cast<int>(v_uint8::nlanes);
+    constexpr int num_vectors = 12;
+
+    if (length < 3 * nlanes)
+        return 0;
+
+    int x = 0;
+
+    v_float32 vectors[num_vectors];
+
+    for (;;)
+    {
+        for (; x <= length - 3 * nlanes; x += 3 * nlanes)
+        {
+            for (int i = 0; i < num_vectors; ++i)
+            {
+                vectors[i] = v_load_f32(in + x + i * nlanes / 4);
```

Comment:
No need to perform hand-made registers spilling. Compilers are smart enough and can do that for you if necessary (moreover AVX512 has up to 32 vector registers)

This data is:
- loaded once
- used once

Move data loading to corresponding places.

Reply:
@alalek Could you please clarify what you mean under hand-made registers spilling. If you mean v_load_f32(), it isn't hand-made registers spilling. It is just an overloaded function for ease of writing templates.

Reply:
1. Data is loaded from the memory.
2. On the same line data is stored back to the memory.
3. Data re-loaded later once again for processing.

Do you see here redundant steps?

P.S. No need to load all 12 SIMD vectors at once. Load data on demand.

Reply:
> P.S. No need to load all 12 SIMD vectors at once. Load data on demand.

It's necessary because of specificities of the algorithm.

Reply:
@terfendail @rgarnov please take a look.

> 1. Data is loaded from the memory.
> 2. On the same line data is stored back to the memory.
> 3. Data re-loaded later once again for processing.
> 
> Do you see here redundant steps?
> 
> P.S. No need to load all 12 SIMD vectors at once. Load data on demand.

Reply:
> No need to perform hand-made registers spilling. Compilers are smart enough and can do that for you if necessary (moreover AVX512 has up to 32 vector registers)
> 
> This data is:
> 
> * loaded once
> * used once
> 
> Move data loading to corresponding places.

Reworked.